src/snowflake/snowpark/dataframe.py

```diff
@@ -520,14 +565,30 @@ class DataFrame:
             3  4.0  6  8
     """
 
+    @publicapi
     def __init__(
         self,
         session: Optional["snowflake.snowpark.Session"] = None,
         plan: Optional[LogicalPlan] = None,
         is_cached: bool = False,
+        _ast_stmt: Optional[proto.Assign] = None,
+        _emit_ast: bool = True,
     ) -> None:
+        """
+        :param int _ast_stmt: The AST Assign atom corresponding to this dataframe value. We track its assigned ID in the
+                             slot self._ast_id. This allows this value to be referred to symbolically when it's
+                             referenced in subsequent dataframe expressions.
+        """
         self._session = session
-        self._plan = self._session._analyzer.resolve(plan)
+        self._ast_id = None
+        if _emit_ast:
+            self._ast_id = _ast_stmt.var_id.bitfield1 if _ast_stmt is not None else None
+
+        if plan is not None:
+            self._plan = self._session._analyzer.resolve(plan)
+        else:
+            self._plan = None
```

Comment:
Why does the AST variant replace the original code with the `plan is not None` check?

Reply:
good question, is this a merge gone wrong? What would be the correct code here?

Reply:
It looks like the change is intentional, and came in with one of Arthur's changes in July. Let's leave it as is.

Reply:
I think pointed this out in one of the original merge PRs for a similar change in dataframe.py and the answer was that it was a bug fix

https://github.com/snowflakedb/snowpark-python/pull/2136#discussion_r1777733248